Update the HDU.fileinfo() behavior in append mode.

hannahrawr opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request updates the HDU.fileinfo() behavior in append mode. Upon appending (and flushing) a HDU to a HDUList opened with mode="append", both fileinfo() methods for the HDU and HDUList now return a dictionary with hdrLoc, datLoc, and datSpan populated.

This is a simple change, where we update the HDU._file reference correctly upon a write, and we update the fileinfo() methods to check and return the required information. Also added a test case for the expected behavior.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #19089 

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

hannahrawr committed:
Update the HDU.fileinfo() behavior in append mode.

hannahrawr committed:
Add a test case for HDUList fileinfo appending

hannahrawr committed:
Rename changelog file to align with PR

hannahrawr committed:
Merge branch 'main' into main

saimn reviewed (commented):
Thanks for the PR, it looks almost good but a few changes are needed (see below).

hannahrawr committed:
Update self._file assignment for HDUList in append case

hannahrawr committed:
Update tests to align with updated append behavior for HDUList

These changes include making the summary formatting consistency between the data loaded and unloaded cases, and updating the tests to correctly append to the HDUList

hannahrawr commented on astropy/io/fits/hdu/base.py:
```diff
@@ -705,6 +705,7 @@ def _writeto_internal(self, fileobj, inplace, copy):
         elif copy:
             datsize = self._writedata_direct_copy(fileobj)
 
+        self._file = fileobj
```
Updated with the self._file assignment above

hannahrawr commented on astropy/io/fits/hdu/table.py:
```diff
@@ -561,7 +561,8 @@ def _summary(self):
 
             ncols = len(self.columns)
-            format = self.columns.formats
+            format = ", ".join(self.columns.formats)
```
I noticed that the formatting between the data loaded and unloaded cases were inconsistent while I was testing it. I've updated it here.

hannahrawr commented on astropy/io/fits/tests/test_hdulist.py:
```diff
@@ -154,7 +183,7 @@ def test_append_extension_to_non_empty_list(self):
 
         with fits.open(self.data("tb.fits")) as hdul:
-            hdul.append(hdul[1])
+            hdul.append(hdul[1].copy())
```
These tests ended up failing after I made the above change, and I think the cause was due to the case where when hdul[1] was being written, the `._file` reference is then updated for _both_ the original and appended HDUs, which subsequently causes an issue when performing the `_writedata`. Making it a copy then sidesteps this issue.

hannahrawr committed:
Merge branch 'main' into main
